DB-183 Added streams: CustomFieldSet and CustomFieldValues for all subjects

adrien-mg opened the pull request:
Added streams: 
* custom_field_sets (http://developer.mavenlink.com/custom_field_sets/), 
* custom_field_values streams for subject: WorkspaceGroup, 
* custom_field_values streams for subject: Story, 
* custom_field_values streams for subject: User (http://developer.mavenlink.com/custom_field_values/)

Curled the endpoints as tests:

```
$ curl -H "Authorization: Bearer <MAVENLINK_TOKEN>" "https://api.mavenlink.com/api/v1/custom_field_sets.json" > custom_fields_sets
$ curl -H "Authorization: Bearer <MAVENLINK_TOKEN>" "https://api.mavenlink.com/api/v1/custom_field_values.json?subject_type=WorkspaceGroup" > workspace_group_custom_field_values
$ curl -H "Authorization: Bearer <MAVENLINK_TOKEN>" "https://api.mavenlink.com/api/v1/custom_field_values.json?subject_type=Story" > story_custom_field_values
$ curl -H "Authorization: Bearer <MAVENLINK_TOKEN>" "https://api.mavenlink.com/api/v1/custom_field_sets.json" > custom_fields_sets
```

adrien-mg committed:
Added streams: custom_field_sets (http://developer.mavenlink.com/custom_field_sets/), and custom_field_values streams for subjects: WorkspaceGroup, Story, User (http://developer.mavenlink.com/custom_field_values/)

davelush-mg commented on tap_mavenlink/streams/custom_field_sets.py:
```diff
@@ -0,0 +1,19 @@
+from tap_mavenlink.streams.base import BaseStream
+import singer
+
+LOGGER = singer.get_logger()  # noqa
```
What does `# noqa` mean on these?

adrien-mg replied:
`# noqa` usually means that linters should ignore that line, although we're not running linters as part of any checks right now :shrug: In this case it's a bit of a copy-paste survivor from other files / lines.

davelush-mg replied:
Nice... Learn something new :D

graham-jackson replied:
+1

davelush-mg commented:
What happens if the custom field value is not a `Story`, `User` or `WorkspaceGroup`? Does it fall into the generic version?

adrien-mg commented:
> What happens if the custom field value is not a Story, User or WorkspaceGroup? Does it fall into the generic version?

@davelush-mg that would return:

```
$ curl -H "Authorization: Bearer <MAVENLINK_TOKEN>" "https://api.mavenlink.com/api/v1/custom_field_values.json?subject_type=Wrong" > wrong_custom_field_values
$ cat wrong_custom_field_values
{"errors":[{"type":"system","message":"Invalid custom field subject type"}]}
```

adrien-mg commented:
Trying to run the tap in `discovery` mode I get the error:

```
CRITICAL 'NoneType' object has no attribute 'to_dict'
Traceback (most recent call last):
  File "tap_mavenlink/__init__.py", line 57, in <module>
    main()
  File "/home/adrien/code/mindgym/data-tap-mavenlink/.venv/test/lib/python3.7/site-packages/singer/utils.py", line 192, in wrapped
    return fnc(*args, **kwargs)
  File "tap_mavenlink/__init__.py", line 46, in main
    args = validate_args(args)
  File "/home/adrien/code/mindgym/data-tap-mavenlink/.venv/test/lib/python3.7/site-packages/tap_mavenlink/utils.py", line 33, in validate_args
    streams = args.catalog.to_dict()['streams']
AttributeError: 'NoneType' object has no attribute 'to_dict'
```

... which I trace back to the file `tap_mavenlink/utils.py` that we introduced in the last PR :speak_no_evil: So I will provide a fix for that in further commits.

graham-jackson reviewed (approved):
LGTM

adrien-mg committed:
Updated Makefile SRC and TESTS variable shell find command, added -or flag for SRC and remove -name .json from TESTS

adrien-mg committed:
Merge branch 'master' into db-183-add-custom-fields-streams

adrien-mg commented:
This PR is cruelly missing the JSON schema for the new fields we are adding so there is a commit coming to add those:
* `tap_mavenlink/schemas/custom_field_sets.json`
* `tap_mavenlink/schemas/user_custom_field_values.json`
* `tap_mavenlink/schemas/story_custom_field_values.json`

Additionally as we don't have any CustomFieldValues for the subject `WorkspaceGroup` we are removing the stream altogether from this PR.

Below are example entries for streams: `WorkspaceCustomFieldValues` **(already defined),** `UserCustomFieldValues` and `StoryCustomFieldValues` to help understand the schema. It is **very** similar to the existing schema for the `WorkspaceCustomFieldValues` stream defined at: [`tap_mavenlink/schemas/workspace_custom_field_values.json`](https://github.com/mind-gym/data-tap-mavenlink/blob/master/tap_mavenlink/schemas/workspace_custom_field_values.json).

We can see that apart from the `value` property (alternatively string, or array) the schemas are identical.

* WorkspaceCustomFieldValues **(already defined)**
```
{
  "subject_type": "workspace",
  "subject_id": 14736687,
  "value": "Leadership development",
  "account_id": 3993967,
  "custom_field_name": "Topic / Solution Set",
  "type": "string",
  "display_value": "Leadership development",
  "can_edit": false,
  "created_at": "2017-01-19T05:15:02-08:00",
  "updated_at": "2017-02-01T04:10:59-08:00",
  "custom_field_id": "189767",
  "setter_id": "6542607",
  "id": "58796595"
}
```

* UserCustomFieldValues
```
{
  "subject_type": "user",
  "subject_id": 10970835,
  "value": [
    2204995
  ],
  "account_id": 3993967,
  "custom_field_name": "Location",
  "type": "single",
  "display_value": "US",
  "can_edit": true,
  "created_at": "2019-09-26T05:02:18-07:00",
  "updated_at": "2019-09-26T05:02:18-07:00",
  "custom_field_id": "568165",
  "setter_id": "6080597",
  "id": "291063025"
}
```

* StoryCustomFieldValues
```
{
  "subject_type": "story",
  "subject_id": 176563737,
  "value": [
    384817
  ],
  "account_id": 3993967,
  "custom_field_name": "4. Product type",
  "type": "multi",
  "display_value": "Custom Toolkit (incl Participant)",
  "can_edit": false,
  "created_at": "2016-12-06T08:21:56-08:00",
  "updated_at": "2016-12-06T08:21:56-08:00",
  "custom_field_id": "178807",
  "setter_id": "6080597",
  "id": "51194817"
}
```

adrien-mg commented:
As for the `CustomFieldSets` stream, below is an example record to help map to the associated JSON schema (coming soon as: `tap_mavenlink/schemas/custom_field_sets.json`):

```
{
  "name": "Line Manager",
  "subject_type": "user",
  "created_at": "2016-07-13T08:31:53-07:00",
  "updated_at": "2016-07-13T08:31:53-07:00",
  "id": "75937"
}
```

adrien-mg committed:
Added schema for streams: CustomFieldSets, StoryCustomFieldValues, UserCustomFieldValues. Also added a test to run the tap in discovery mode. Removed WorkspaceGroupCustomFieldValues streams as it is currently empty in our Mavenlink source. Remove *.json from .gitignore,

graham-jackson commented on tap_mavenlink/schemas/user_custom_field_values.json:
```diff
@@ -0,0 +1,82 @@
+{
```
This file looks the same as `story_custom_field_values.json`. 
Part of me thinks, 'Why not try to have one file rather than two?', but the other part of me doesn't mind 🤔 

I'm happy to keep it simple and have two files 👍

adrien-mg replied:
Happy to keep it simple :+1:

graham-jackson commented on tests/test_discover.py:
```diff
@@ -0,0 +1,15 @@
+from tap_mavenlink import main
+
+
+def test_discover_no_catalog(tmp_path, config):
+    sys_args = [
+        'tap-mavenlink',
+        '--config', str(tmp_path / constants.CONFIG_FILENAME),
+        '--discover'
+    ]
+    with patch('sys.argv', sys_args):
+        main()
```
```suggestion
def test_discover_no_catalog(tmp_path, config):
    # Given
    sys_args = [
        'tap-mavenlink',
        '--config', str(tmp_path / constants.CONFIG_FILENAME),
        '--discover'
    ]
    
    # When
    with patch('sys.argv', sys_args):
        main()
        
    # Then
    # Should we prove that we catch some stdout output here?
```

graham-jackson reviewed (approved):
Two minor comments. 

Pre-emptively approving.

adrien-mg committed:
Added expected discovery json file to tests for regression testing

graham-jackson reviewed (approved):
LGTM

adrien-mg committed:
Added JSON Schema to validate catalog (discovery) schema in test_discover file. Removed Tuple Validation from {story,user}_custom_field_values.json schemas as it is currently not supported by the singer-python library, and it will fail when entering the main decorator: @singer.utils.handle_top_exception. Issue was raised at: https://github.com/singer-io/singer-python/issues/145. Also fixed a tiny typo in custom_field_sets.json schema (objects -> object)

adrien-mg committed:
Bumping minor version to include 3 new streams to integrate: custom_field_sets, user_custom_field_values, story_custom_field_values

adrien-mg committed:
Resolving conflict with latest master on new version with minor bump

adrien-mg commented:
The new (should be the last :sweat:) set of changes basically fixes schemas:
* In: `tap_mavenlink/schemas/custom_field_sets.json` we fixed a type: `objects` to `object`
* In: `tap_mavenlink/schemas/user_custom_field_values.json` and `tap_mavenlink/schemas/story_custom_field_values.json` we fixed the `value` property which was previously defined as a tuple of length one, into either a `string` or a simple `array` using the `anyOf` syntax.

It also adds a JSON schema for the catalog that we use in the `tests/test_discover.py` file to validate that the discovered catalog hasn't regressed and that it further matches an expected JSON schema at: tests/catalog-schema.json`

adrien-mg committed:
Making dependency upon `jsonschema` library used in test explicit in the setup.py file. Noting that it was previously implicitely imported by the `tap_framework` dependency

graham-jackson commented on tests/catalog-schema.json:
```diff
@@ -0,0 +1,59 @@
+{
```
I need help with what this file is for/the file name:

I think a tap can have a `catalog.json`.
The catalog can have multiple streams, and each stream can have a schema.

...Is there a catalog schema too?

graham-jackson replied:
Ah, I think I get it: take the catalog, and use jsonschema to get the structure of it. ...That's what's in this file...

👍

adrien-mg replied:
Yes :+1:

graham-jackson commented on tests/test_discover.py:
```diff
@@ -6,6 +8,7 @@
 
 
 EXPECTED_DISCOVERY_CONTENT = './tests/discovery.json'
+EXPECTED_DISCOVERY_SCHEMA_CONTENT = './tests/catalog-schema.json'
```
I prefer 'catalog' rather than 'discovery' in variable and file names. 

...Because it makes me think that the 'discovery.json' is something different to 'catalog.json', which I don't think is true.

Not a big issue. Happy to leave as is if you prefer it that way 🙇

adrien-mg replied:
Renamed :+1:

graham-jackson reviewed (approved):
One minor comment. Pre-approving.

adrien-mg committed:
Renamed `discovery` variable names to something more like `catalog`, fixed the tests to match `dict` objects instead of `str` objects.
